### PR TITLE
Git hooks libms

### PR DESCRIPTION
**Pull request by lindot11**
- The pre-commit framework can replace the existing git hooks smoothly
  - Benefits: code quality improvement in terms of consistency, automation, maintainability
  - Hooks are installed in Gits default hook path ./git/hooks/ for the automated checks to work
  - If this replacement is approved, the .git-hooks directory and configure-git-hooks.sh script can be removed
  - The githook_test.sh, githook_test_log.txt and githook_run_log.txt shall be removed before merging to upstream
- The hooks have successfully been tested
  - Logs can be seen [here](https://github.com/lindot11/DTaaS/blob/feature/distributed-demo/githook_test_log.txt)
- A new git hooks script was created and the [/docs/developer/index.md](https://github.com/lindot11/DTaaS/blob/feature/distributed-demo/docs/developer/index.md) have been updated
- A hook for testing the lib on pre-push have also been added to the [.pre-commit-config.yaml](https://github.com/INTO-CPS-Association/DTaaS/assets/92531748/2e52d4a8-7c6e-48d5-95b9-d2d5cb4064f3)

  - The hook is commented-out because when run the test fails and exits
- Git hook corrections
  - We have accepted the format and syntax corrections from the git hooks that we assume are fine
  - This [log](https://github.com/lindot11/DTaaS/blob/git-hooks-libms/githook_run_log.txt) shows the suggested corrections that we have not accepted and raises the following questions
    - What should the markdown line length be configured to?
    - What should we title the header for this [README](https://github.com/lindot11/DTaaS/blob/git-hooks-libms/client/public/static/font/README.md)?
    - Can we remove this [README](https://github.com/lindot11/DTaaS/blob/git-hooks-libms/servers/lib/test/data/user2/tools/README.md)?

- Try the hooks manually
	- Setup
		- bash script/setup-git-hooks.sh
	- Run
		- pre-commit run --hook-stage pre-commit
		- pre-commit run --hook-stage pre-push

### CODE REVIEW

File: servers/execution/runner/README.md [hunk 1]
```diff
@@ -81,11 +81,12 @@ Access to the service on network is available at `http://<ip or hostname>:3000/`
...
 
-| REST API Route | Return Value | Comment |
-|:---|:---|:---|
-| localhost:3000/phase | [ hello ] | The array get appended with each invocation. All the elements of are _array_. |
-| localhost:3000/lifecycle/phase | _true_ | Always returns _true_ |
-|||
+<!-- markdownlint-disable MD013 -->
+| REST API Route                 | Return Value | Comment |
+| :----------------------------- | :----------- | :------ |
+| localhost:3000/phase           | [ hello ]    | The array get appended with each invocation. All the elements of are _array_. |
```
[the remark this answers is not in the document]
Comment (lindot11) [reply servers/execution/runner/README.md:87]: ignore this line length

**Comment by lindot11** [comment]
Finish revisions after feedback
- the code climate detects line_length issues on code-blocks which I assume should be ignored

Summary of the 52 files changed:
- .git-hooks replaced by pre-commit framework (.pre-commit-config.yaml)
- adjusted markdownlint node.js implementation to follow the codeclimates mdl ruby rules
  - with markdownlint node.js we can disable markdownlint rules on specific lines that we don't want to change
- added prettier to client and lib
- accepted prettier format corrections which does not affect codecov

File: servers/lib/package.json [hunk 2]
```diff
@@ -56,7 +57,7 @@
     "eslint-plugin-prettier": "^5.0.0",
     "graphql-scalars": "^1.22.2",
     "jest": "29.7.0",
-    "prettier": "^3.0.3",
```
Comment (prasadtalasila) [inline servers/lib/package.json:59]: please skip the package upgrades in this PR.

File: servers/lib/yarn.lock [hunk 3]
```diff
@@ -5053,10 +5053,10 @@ prettier-linter-helpers@^1.0.0:
   dependencies:
     fast-diff "^1.1.2"
```
Comment (prasadtalasila) [inline servers/lib/yarn.lock:5055]: please restore the lock file as well.

**Comment by prasadtalasila** [comment]
@lindot11, A few more suggestions:

1. In docs/developer/index.md, after this sentence: "The git-hooks will ensure that your commits are formatted correctly and that the tests pass before you push the commits to remote repositories."
  add these two commands and their explanation
  ```  
    pre-commit run --hook-stage pre-commit
    pre-commit run --hook-stage pre-push
  ```

2. Please add shellcheck, json and yaml plugins as well.
3. The files in docs/, deploy/, script/ and ssl/ directories need to be linted by eslint, shellcheck, json and yaml plugins. This step is to be part of pre-commit stage hook.
4. Please rebase the PR and squash the commits. There are almost 250+ commits in this PR.

Thanks.

**Comment by lindot11** [comment]
After running pre-commit run -a, the markdownlint checks suggest many corrections for the files. Should i solve it in this in this PR?

```bash
markdownlint.............................................................Failed
- hook id: markdownlint
- exit code: 1

docs/admin/guides/add_user.md:10:1 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/guides/add_user.md:10:15 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/guides/add_user.md:14:1 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/guides/add_user.md:14:28 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/guides/add_user.md:34:1 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/guides/add_user.md:34:32 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/guides/add_user.md:43:1 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/guides/add_user.md:43:31 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/guides/add_user.md:79:1 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/guides/add_user.md:79:26 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/overview.md:27:50 MD034/no-bare-urls Bare URL used [Context: "https://foo.com"]
docs/admin/overview.md:28:14 MD034/no-bare-urls Bare URL used [Context: "http://foo.com"]
docs/admin/vagrant/base-box.md:106:62 MD047/single-trailing-newline Files should end with a single newline character
docs/user/digital-twins/lifecycle.md:31:81 MD013/line-length Line length [Expected: 80; Actual: 138]
docs/user/digital-twins/lifecycle.md:32:81 MD013/line-length Line length [Expected: 80; Actual: 122]
docs/user/digital-twins/lifecycle.md:33:81 MD013/line-length Line length [Expected: 80; Actual: 153]
docs/user/digital-twins/lifecycle.md:34:81 MD013/line-length Line length [Expected: 80; Actual: 122]
docs/user/examples/drobotti-rmqfmu/README.md:70:81 MD013/line-length Line length [Expected: 80; Actual: 96]
docs/user/examples/water_tank_swap/README.md:36:81 MD013/line-length Line length [Expected: 80; Actual: 85]
docs/user/examples/water_tank_swap/README.md:38:81 MD013/line-length Line length [Expected: 80; Actual: 105]
docs/user/features.md:35:81 MD013/line-length Line length [Expected: 80; Actual: 163]
docs/user/features.md:36:81 MD013/line-length Line length [Expected: 80; Actual: 86]
docs/user/features.md:37:81 MD013/line-length Line length [Expected: 80; Actual: 126]
docs/user/features.md:38:81 MD013/line-length Line length [Expected: 80; Actual: 92]
docs/user/features.md:49:81 MD013/line-length Line length [Expected: 80; Actual: 218]
docs/user/features.md:50:81 MD013/line-length Line length [Expected: 80; Actual: 88]
docs/user/features.md:52:81 MD013/line-length Line length [Expected: 80; Actual: 109]
docs/user/website/index.md:127:5 MD037/no-space-in-emphasis Spaces inside emphasis markers [Context: "* d"]
docs/user/website/index.md:129:5 MD037/no-space-in-emphasis Spaces inside emphasis markers [Context: "* m"]
deploy/README.md:108:47 MD034/no-bare-urls Bare URL used [Context: "https://gitlab.foo.com/"]
deploy/README.md:110:43 MD034/no-bare-urls Bare URL used [Context: "https://foo.com/Library"]
deploy/README.md:111:81 MD013/line-length Line length [Expected: 80; Actual: 87]
deploy/README.md:161:68 MD034/no-bare-urls Bare URL used [Context: "https://foo.com"]
deploy/vagrant/two-machine/README.md:44:48 MD034/no-bare-urls Bare URL used [Context: "WWW"]
docs/admin/guides/add_service.md:21:81 MD013/line-length Line length [Expected: 80; Actual: 99]
docs/admin/guides/add_service.md:22:81 MD013/line-length Line length [Expected: 80; Actual: 99]
docs/admin/guides/add_service.md:23:81 MD013/line-length Line length [Expected: 80; Actual: 99]
docs/admin/guides/add_service.md:24:81 MD013/line-length Line length [Expected: 80; Actual: 99]
docs/admin/guides/add_service.md:25:81 MD013/line-length Line length [Expected: 80; Actual: 99]
docs/admin/guides/add_service.md:26:81 MD013/line-length Line length [Expected: 80; Actual: 99]
docs/FAQ.md:57:81 MD013/line-length Line length [Expected: 80; Actual: 104]
docs/FAQ.md:68:81 MD013/line-length Line length [Expected: 80; Actual: 240]
docs/FAQ.md:72:81 MD013/line-length Line length [Expected: 80; Actual: 113]
docs/FAQ.md:93:81 MD013/line-length Line length [Expected: 80; Actual: 172]
docs/FAQ.md:104:81 MD013/line-length Line length [Expected: 80; Actual: 164]
docs/FAQ.md:123:81 MD013/line-length Line length [Expected: 80; Actual: 120]
docs/FAQ.md:135:81 MD013/line-length Line length [Expected: 80; Actual: 191]
docs/FAQ.md:149:81 MD013/line-length Line length [Expected: 80; Actual: 155]
docs/FAQ.md:155:81 MD013/line-length Line length [Expected: 80; Actual: 131]
docs/FAQ.md:171:81 MD013/line-length Line length [Expected: 80; Actual: 102]
docs/FAQ.md:178:81 MD013/line-length Line length [Expected: 80; Actual: 100]
docs/FAQ.md:262:81 MD013/line-length Line length [Expected: 80; Actual: 151]
LICENSE.md:23:49 MD034/no-bare-urls Bare URL used [Context: "http://www.into-cps.org"]
LICENSE.md:26:10 MD034/no-bare-urls Bare URL used [Context: "http://www.gnu.org/copyleft/gp..."]
LICENSE.md:294:8 MD034/no-bare-urls Bare URL used [Context: "http://www.eclipse.org/legal/e..."]
LICENSE.md:298:8 MD034/no-bare-urls Bare URL used [Context: "http://www.gnu.org/copyleft/gp..."]
docs/admin/client/auth.md:45:81 MD013/line-length Line length [Expected: 80; Actual: 90]
docs/admin/client/auth.md:92:81 MD013/line-length Line length [Expected: 80; Actual: 92]
docs/admin/client/auth.md:94:81 MD013/line-length Line length [Expected: 80; Actual: 123]
docs/admin/client/auth.md:94:3 MD034/no-bare-urls Bare URL used [Context: "https://foo.com/au"]
docs/admin/client/auth.md:94:78 MD034/no-bare-urls Bare URL used [Context: "https://foo.com/au"]
docs/admin/client/auth.md:94:49 MD034/no-bare-urls Bare URL used [Context: "https://foo.com/au/Library"]
docs/admin/client/auth.md:94:24 MD034/no-bare-urls Bare URL used [Context: "https://foo.gitlab.com"]
docs/admin/client/auth.md:95:81 MD013/line-length Line length [Expected: 80; Actual: 125]
docs/admin/client/auth.md:95:3 MD034/no-bare-urls Bare URL used [Context: "https://foo.com/acme"]
docs/admin/client/auth.md:95:80 MD034/no-bare-urls Bare URL used [Context: "https://foo.com/au"]
docs/admin/client/auth.md:95:51 MD034/no-bare-urls Bare URL used [Context: "https://foo.com/au/Library"]
docs/admin/client/auth.md:95:26 MD034/no-bare-urls Bare URL used [Context: "https://foo.gitlab.com"]
docs/admin/client/auth.md:96:81 MD013/line-length Line length [Expected: 80; Actual: 124]
docs/admin/client/auth.md:96:79 MD034/no-bare-urls Bare URL used [Context: "https://foo.com/au"]
docs/admin/client/auth.md:96:50 MD034/no-bare-urls Bare URL used [Context: "https://foo.com/au/Library"]
docs/admin/client/auth.md:96:3 MD034/no-bare-urls Bare URL used [Context: "https://foo.com/bar"]
docs/admin/client/auth.md:96:25 MD034/no-bare-urls Bare URL used [Context: "https://foo.gitlab.com"]
docs/admin/client/auth.md:112:3 MD047/single-trailing-newline Files should end with a single newline character
docs/admin/host.md:119:81 MD013/line-length Line length [Expected: 80; Actual: 108]
docs/admin/host.md:120:81 MD013/line-length Line length [Expected: 80; Actual: 108]
docs/admin/host.md:121:81 MD013/line-length Line length [Expected: 80; Actual: 108]
docs/admin/host.md:121:59 MD034/no-bare-urls Bare URL used [Context: "https://gitlab.foo.com/"]
docs/admin/host.md:123:81 MD013/line-length Line length [Expected: 80; Actual: 108]
docs/admin/host.md:123:59 MD034/no-bare-urls Bare URL used [Context: "https://foo.com/Library"]
docs/admin/host.md:124:81 MD013/line-length Line length [Expected: 80; Actual: 108]
docs/admin/host.md:179:68 MD034/no-bare-urls Bare URL used [Context: "https://foo.com"]
docs/admin/servers/lib/LIB-MS.md:106:52 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/servers/lib/LIB-MS.md:106:59 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/servers/lib/LIB-MS.md:108:20 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/servers/lib/LIB-MS.md:108:27 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/servers/lib/LIB-MS.md:115:8 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/servers/lib/LIB-MS.md:115:15 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
docs/admin/trial.md:32:81 MD013/line-length Line length [Expected: 80; Actual: 108]
docs/admin/trial.md:33:81 MD013/line-length Line length [Expected: 80; Actual: 108]
docs/admin/trial.md:34:81 MD013/line-length Line length [Expected: 80; Actual: 108]
docs/admin/trial.md:34:59 MD034/no-bare-urls Bare URL used [Context: "https://gitlab.foo.com/"]
docs/admin/trial.md:36:81 MD013/line-length Line length [Expected: 80; Actual: 108]
docs/admin/trial.md:36:59 MD034/no-bare-urls Bare URL used [Context: "https://foo.com/Library"]
docs/admin/trial.md:37:81 MD013/line-length Line length [Expected: 80; Actual: 108]
docs/LICENSE.md:23:49 MD034/no-bare-urls Bare URL used [Context: "http://www.into-cps.org"]
docs/LICENSE.md:25:42 MD034/no-bare-urls Bare URL used [Context: "http://www.gnu.org/copyleft/gp..."]
docs/LICENSE.md:294:8 MD034/no-bare-urls Bare URL used [Context: "http://www.eclipse.org/legal/e..."]
docs/LICENSE.md:299:8 MD034/no-bare-urls Bare URL used [Context: "http://www.gnu.org/copyleft/gp..."]
docs/user/servers/lib/LIB-MS.md:41:26 MD034/no-bare-urls Bare URL used [Context: "https://foo.com/lib"]
docs/user/servers/lib/LIB-MS.md:45 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/user/servers/lib/LIB-MS.md:75 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/user/servers/lib/LIB-MS.md:134 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/user/servers/lib/LIB-MS.md:148 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/user/servers/lib/LIB-MS.md:168 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/user/servers/lib/LIB-MS.md:186 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/user/servers/lib/LIB-MS.md:208 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
docs/user/servers/lib/LIB-MS.md:220 MD046/code-block-style Code block style [Expected: fenced; Actual: indented]
deploy/services/README.md:16:40 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
deploy/services/README.md:16:54 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
deploy/services/README.md:17:32 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
deploy/services/README.md:17:51 MD050/strong-style Strong style should be consistent [Expected: asterisk; Actual: underscore]
servers/lib/test/data/user2/tools/README.md:1 MD041/first-line-heading/first-line-h1 First line in a file should be a top-level heading [Context: "content123"]
```

**Comment by prasadtalasila** [comment]
@lindot11
The git commit does not work if pre-commit hook fails. If these linting issues exist, then the commits will not go forward right?
Is there a way to let pre-commit exit with success status (exit 0 status) even if the markdown linter throws up errors?
This is equivalent to making a pre-commit plugin produce warnings for violations instead of errors.

**Comment by lindot11** [comment]
I could not find args or configs settings that could fix this issue, so I made a script.

The [script](https://github.com/lindot11/DTaaS/blob/git-hooks-libms/script/markdownlint-hook.sh) wraps the markdownlint hook and outputs the linting issues, but exits with code 0. 
Then configured the markdownlint hook verbose setting to true, so that it will always show the detected linting issues when exited 0.

**Comment by prasadtalasila** [comment]
Merged via PR #344

**Pull request closed without merging**